src/server/worker/pipeline/common.go

```diff
@@ -64,94 +65,61 @@ func ReceiveSpout(
 	pipelineInfo *pps.PipelineInfo,
 	logger logs.TaggedLogger,
 ) (retErr1 error) {
-	return backoff.RetryUntilCancel(ctx, func() error {
-		repo := pipelineInfo.Pipeline.Name
-		// open a read connection to the /pfs/out named pipe
+	// Check to see if this spout is the latest version of this spout by seeing if its spec commit has any children.
+	spec, err := pachClient.InspectCommit(ppsconsts.SpecRepo, pipelineInfo.SpecCommit.ID)
+	if err != nil && !errutil.IsNotFoundError(err) {
+		return err
+	}
+	if spec != nil && len(spec.ChildCommits) != 0 {
+		return errors.New("outdated spout, now shutting down")
+	}
+	repo := pipelineInfo.Pipeline.Name
+	return backoff.RetryUntilCancel(ctx, func() (retErr error) {
```

Comment:
@jdoliner I'm thinking about removing this retry because I think we would only hit this when a serious error occurs. Particularly errors where we shouldn't retry because we could end up with corrupt data (issues reading the tar streams).

Reply:
That seems reasonable to me, when we encounter an error here would we wind up recreating the pipe and restarting the user's code? I think that's the only way to handle this error case.

Reply:
Yeah, I think that is the only way to handle the situation when we get out of sync with the writer. The alternative is to silently lose and / or corrupt data (not exactly silent because of the logs, but it won't be obvious that data was lost and / or corrupted without catching the log message).